web/src/main/java/se/inera/intyg/webcert/web/integration/interactions/listcertificatesforcarewithqa/ListCertificatesForCareWithQAResponderImpl.java

```diff
@@ -20,72 +20,90 @@
+        final var response = new ListCertificatesForCareWithQAResponseType();
+        final var start = System.currentTimeMillis();
+        log.info("Started processing request: PersonId: '{}' - CareProviderId '{}' - UnitIds '{}'",
+            HashUtility.hash(intygWithNotificationsRequest.getPersonnummer().getPersonnummer()),
+            intygWithNotificationsRequest.getVardgivarId(),
+            intygWithNotificationsRequest.getEnhetId()
+        );
+        try {
+            final var notifications = notificationService.findNotifications(intygWithNotificationsRequest);
+            final var listItemsFromCS = getCertificatesWithQAFromCertificateService.get(notifications);
+            final var intygWithNotifications = intygService.listCertificatesForCareWithQA(
+                notifications.stream()
+                    .filter(removeNotificationsRelatedToCertificatesFromCertificateService(listItemsFromCS))
+                    .collect(Collectors.toList())
+            );
 
-        if (!request.getEnhetsId().isEmpty()) {
-            builder = builder.setEnhetId(request.getEnhetsId().stream().map(HsaId::getExtension).collect(Collectors.toList()));
-        }
-        if (request.getVardgivarId() != null) {
-            builder = builder.setVardgivarId(request.getVardgivarId().getExtension());
-        }
-        if (request.getFromTidpunkt() != null) {
-            builder = builder.setStartDate(request.getFromTidpunkt());
-        }
-        if (request.getTomTidpunkt() != null) {
-            builder = builder.setEndDate(request.getTomTidpunkt());
-        }
+            List list = new List();
+            list.getItem().addAll(buildListItemsForWC(intygWithNotifications));
+            list.getItem().addAll(listItemsFromCS);
+            response.setList(list);
 
-        final var intygWithNotificationsRequest = builder.build();
-        final var listItemsFromCS = getPatientCertificatesWithQAFromCertificateService.get(intygWithNotificationsRequest);
-        java.util.List<IntygWithNotificationsResponse> intygWithNotifications = intygService.listCertificatesForCareWithQA(
-            intygWithNotificationsRequest,
-            listItemsFromCS.stream()
-                .map(ListItem::getIntyg)
-                .map(Intyg::getIntygsId)
-                .map(IIType::getExtension)
-                .collect(Collectors.toList())
-        );
+            return response;
+        } finally {
+            log.info(
+                "Request processing completed. PersonId: '{}' CareProviderId '{}' UnitIds '{}'."
+                    + " Returning '{}' number of certificates. Elapsed time: '{}' seconds",
+                HashUtility.hash(intygWithNotificationsRequest.getPersonnummer().getPersonnummer()),
+                intygWithNotificationsRequest.getVardgivarId(),
+                intygWithNotificationsRequest.getEnhetId(),
+                response.getList().getItem().size(),
```

Comment:
This row will throw an NPE if any exception has been thrown between the try / finally clause. Just add a null check on the response.getList() it should be fine.